[Style] Convert border-radius to use strong style types

weinig opened the pull request:
#### 0104cc3c5a1172346ea4510603a8de0556803c66
<pre>
[Style] Convert border-radius to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=294982">https://bugs.webkit.org/show_bug.cgi?id=294982</a>

Reviewed by Darin Adler.

Converts border radius in BorderData to use a strong style type, taking
advantage of the existing work done for the radii of CSS shapes, tweaking
it a bit by adopting RectCorners directly.

To allow forward declaring style primitive types, we now have
them declared in a new minimal &quot;StylePrimitiveNumeric+Forward.h&quot;
which can be included in headers without bloat.

Fixes issue with path shrink-wrapping code that assumed border-radius
was always fixed.

* LayoutTests/fast/css/shrink-wrap-border-radius-with-calc-crash-expected.txt: Added.
* LayoutTests/fast/css/shrink-wrap-border-radius-with-calc-crash.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/platform/graphics/PathUtilities.cpp:
* Source/WebCore/platform/graphics/PathUtilities.h:
* Source/WebCore/rendering/BorderPainter.cpp:
* Source/WebCore/rendering/BorderShape.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
* Source/WebCore/rendering/shapes/LayoutShape.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/borders/StyleBorderRadius.cpp:
* Source/WebCore/style/values/borders/StyleBorderRadius.h:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric+Forward.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ios/WKTapHighlightView.mm:
* Source/WebKitLegacy/ios/Misc/WebUIKitSupport.mm:

Canonical link: <a href="https://commits.webkit.org/296770@main">https://commits.webkit.org/296770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d78adcfdc9f69d6d21fb91a2fc1c333a7b546a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109538 "22 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29196 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19625 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114740 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59741 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111501 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29874 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37784 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83258 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23803 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98664 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63718 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23182 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16807 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59357 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93177 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16847 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117853 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36577 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27084 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92269 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36950 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94924 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92086 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23449 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37024 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14771 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32375 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36471 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41942 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36130 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39474 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37841 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/PathUtilities.cpp:
```diff
@@ -471,13 +470,14 @@ static FloatRoundedRect::Radii adjustedRadiiForHuggingCurve(const FloatSize& top
+    radii.setTopRight(adjustedRadius(inputRadii.topRight()));
+    radii.setBottomRight(adjustedRadius(inputRadii.bottomRight()));
+    radii.setBottomLeft(adjustedRadius(inputRadii.bottomLeft()));
     return radii;
```
Kind of wish this was construction rather than done with setters.

weinig replied:
Yeah, I should fix it.

darinadler commented on Source/WebCore/platform/graphics/PathUtilities.h:
```diff
@@ -27,20 +27,25 @@
     WEBCORE_EXPORT static Vector<Path> pathsWithShrinkWrappedRects(const Vector<FloatRect>& rects, float radius);
 
-    static Path pathWithShrinkWrappedRectsForOutline(const Vector<FloatRect>&, const BorderData&, float outlineOffset, WritingMode, float deviceScaleFactor);
+    // FIXME: This is a layering violation as it depends on types outside of `platform`.
```
Sounds worth fixing. Good call.

darinadler commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1490,7 +1491,13 @@ bool RenderElement::repaintAfterLayoutIfNeeded(SingleThreadWeakPtr<const RenderL
+                        Style::evaluate(style.borderTopRightRadius().width(), borderBoxWidth),
+                        Style::evaluate(style.borderBottomRightRadius().width(), borderBoxWidth)
+                    )
+                );
```
Can't this be done with a single call to `std::max` that takes three arguments?

weinig replied:
Oh, yeah, it totally can.

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumeric+Forward.h:
```diff
@@ -24,33 +24,36 @@
+// MARK: Integer Primitive
 
-template<typename CSSType> struct CSSValueConversions;
+template<CSS::Range R = CSS::All, typename V = int> struct Integer;
```
```suggestion
template<CSS::Range = CSS::All, typename = int> struct Integer;
```

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumeric+Forward.h:
```diff
@@ -24,33 +24,36 @@
+// MARK: Percentage Primitive
 
-} // namespace CSS
+template<CSS::Range R = CSS::All, typename V = double> struct Percentage;
```
```suggestion
template<CSS::Range = CSS::All, typename = double> struct Percentage;
```

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumeric+Forward.h:
```diff
@@ -24,33 +24,36 @@
+template<CSS::Range R = CSS::All, typename V = double> struct Time;
+template<CSS::Range R = CSS::All, typename V = double> struct Frequency;
+template<CSS::Range R = CSS::Nonnegative, typename V = double> struct Resolution;
+template<CSS::Range R = CSS::All, typename V = double> struct Flex;
```
```suggestion
template<CSS::Range = CSS::All, typename = double> struct Angle;
template<CSS::Range = CSS::All, typename = float> struct Length;
template<CSS::Range = CSS::All, typename = double> struct Time;
template<CSS::Range = CSS::All, typename = double> struct Frequency;
template<CSS::Range = CSS::Nonnegative, typename = double> struct Resolution;
template<CSS::Range = CSS::All, typename = double> struct Flex;
```

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumeric+Forward.h:
```diff
@@ -24,33 +24,36 @@
+// MARK: Dimension + Percentage Primitives
+
+template<CSS::Range R = CSS::All, typename V = float> struct AnglePercentage;
+template<CSS::Range R = CSS::All, typename V = float> struct LengthPercentage;
```
```suggestion
template<CSS::Range = CSS::All, typename = float> struct AnglePercentage;
template<CSS::Range = CSS::All, typename = float> struct LengthPercentage;
```

weinig committed:
[Style] Convert border-radius to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=294982

Reviewed by Darin Adler.

Converts border radius in BorderData to use a strong style type, taking
advantage of the existing work done for the radii of CSS shapes, tweaking
it a bit by adopting RectCorners directly.

To allow forward declaring style primitive types, we now have
them declared in a new minimal "StylePrimitiveNumeric+Forward.h"
which can be included in headers without bloat.

Fixes issue with path shrink-wrapping code that assumed border-radius
was always fixed.

* LayoutTests/fast/css/shrink-wrap-border-radius-with-calc-crash-expected.txt: Added.
* LayoutTests/fast/css/shrink-wrap-border-radius-with-calc-crash.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/platform/graphics/PathUtilities.cpp:
* Source/WebCore/platform/graphics/PathUtilities.h:
* Source/WebCore/rendering/BorderPainter.cpp:
* Source/WebCore/rendering/BorderShape.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
* Source/WebCore/rendering/shapes/LayoutShape.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/borders/StyleBorderRadius.cpp:
* Source/WebCore/style/values/borders/StyleBorderRadius.h:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric+Forward.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ios/WKTapHighlightView.mm:
* Source/WebKitLegacy/ios/Misc/WebUIKitSupport.mm:

Canonical link: https://commits.webkit.org/296770@main
